### PR TITLE
[CARBONDATA-465] Spark streaming dataframe support

### PR DESCRIPTION
**Pull request by allwefantasy**
* mvn clean verify have already been passed locally.
* No new unit test cases are added
* Tested in streamingpro project.
* Remove kettle clearly
* Fix NLP of loading data with new flow when sometimes spark executor will use  TaskContext.get but the code run in another thread

### CODE REVIEW

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 1]
```diff
@@ -763,10 +763,9 @@ case class LoadTable(
 
 
       val columinar = sqlContext.getConf("carbon.is.columnar.storage", "true").toBoolean
-      val kettleHomePath = CarbonScalaUtil.getKettleHome(sqlContext)
 
       // TODO It will be removed after kettle is removed.
-      val useKettle = options.get("use_kettle") match {
+      val useKettle = options.get("useKettle") match {
```
Comment (ravipesala) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: why changed the string? better avoid changing it, it is used in other places as well

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java [hunk 2]
```diff
@@ -604,4 +606,33 @@ public static boolean isHeaderValid(String tableName, String header,
     }
     return dateformatsHashMap;
   }
+
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: This reflection code is not required. Please check how load happens through dataframe in CarbonDataLoadRDD. I guess you can make use of same code.
And your iterator supposed to extend `InputIterator` of PR33.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala (hunk 1 above)
Comment (allwefantasy) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: DataFrameWriter accept parameter `useKettle` eg.  writer.option("useKettle", "false").option("tempCSV","false").  Maybe we should keep consistent。

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonMetastoreCatalog.scala [hunk 3]
```diff
@@ -307,10 +307,17 @@ class CarbonMetastoreCatalog(hiveContext: HiveContext, val storePath: String,
     if (!FileFactory.isFileExist(schemaMetadataPath, fileType)) {
       FileFactory.mkdirs(schemaMetadataPath, fileType)
     }
+
+    /**
+    * schemaFilePath starts with file:// will not create meta files successfully
+    * while thriftWriter will have no complains.
+    * This will cause some weired error eg. No table found.
+    */
```
Comment (allwefantasy) [inline integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonMetastoreCatalog.scala]: Ok .

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java (hunk 2 above)
Comment (allwefantasy) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: Module carbon-processing do not depends on spark or other computing engine however there are some class need multi-thread to load data which runs as computing engine's task which need get TaskContext using ThreadLocal tech.   

Yes, my first PR is merged from your PR333,but it's not merged to master yet.

Comment (allwefantasy) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: i have removed PR333  from this PR

**Comment by allwefantasy** [comment]
@jackylk  fix confict. Can you review it?

**Comment by ravipesala** [comment]
@allwefantasy  it seems there is a problem with merging/rebasing the PR, other commits has come here. Please fix it.

**Comment by allwefantasy** [comment]
It's weird.  In my local branch, git log shows:

```
commit acdf78a8cba4f7c18cbaaf0fcc1a9e9dc3189068
Merge: 8a21cb7 5ca7218
Author: WilliamZhu <allwefantasy@gmail.com>
Date:   Mon Dec 5 11:30:35 2016 +0800

    Merge branch 'spark-streaming-dataframe-support2' of github.com:allwefantasy/incubator-carbondata into spark-streaming-dataframe-support2

commit 8a21cb715eac50c04b859530ab459ae9b6f226a3
Author: WilliamZhu <allwefantasy@gmail.com>
Date:   Wed Nov 30 21:24:33 2016 +0800

    remove comments on createTableFromThrift and rais jira later

commit 06bc4239a2762a6f27da99982b47e880d6a1be4c
Author: WilliamZhu <allwefantasy@gmail.com>
Date:   Wed Nov 30 00:12:24 2016 +0800

    reset maven-source-plugin

commit 0f042797f54143bd473296bc33650e84d071dd15
Author: WilliamZhu <allwefantasy@gmail.com>
Date:   Tue Nov 29 23:46:42 2016 +0800

    spark streaming dataframe support

commit 70ae82045e461c740cf2ae80c2058160bc9855a9
Merge: e7958b6 fc3f6b3
Author: ravipesala <ravi.pesala@gmail.com>
```

I will try to figure out

**Comment by allwefantasy** [comment]
The commit log shows Changes allwefantasy and others added some commits 5 days ago.  I guess there is no problem @ravipesala

**Comment by allwefantasy** [comment]
Yes It seems not ok...... I will try to figure out how to resolve this.

**Comment by ravipesala** [comment]
Can one of the admins verify this patch?

**Comment by allwefantasy** [comment]
Fix commit log issue and update to latest

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by chenliang613** [comment]
add to whitelist

**Comment by CarbonDataQA** [comment]
Build Success, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/35/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/NewCarbonDataLoadRDD.scala [hunk 4]
```diff
@@ -308,3 +310,155 @@ class NewCarbonDataLoadRDD[K, V](
     }
   }
 }
+
```
Comment (chenliang613) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/NewCarbonDataLoadRDD.scala]: Please add comments  for the new class NewDataFrameLoaderRDD.

Comment (allwefantasy) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/NewCarbonDataLoadRDD.scala]: Comments have been added for NewDataFrameLoaderRDD and NewRddIterator

**Comment by CarbonDataQA** [comment]
Build Success, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/56/

**Comment by CarbonDataQA** [comment]
Build Success, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/57/

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java (hunk 2 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: Please remove this reflection code and try to set from RDD like how it was doing in CarbonDataLoadRDD

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala (hunk 1 above)
Comment (ravipesala) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: I think it does not work if it executes from parser, please make it consistent across all.

**Comment by ravipesala** [comment]
LGTM, Good work.
But few comments like reflection is not required(can set context from initialize from RDD ) are not addressed, we can fix them in another PR.

**Pull request closed without merging**